### PR TITLE
Feature/seb boris clean

### PR DESCRIPTION
**Pull request by sebj101**
Pull request of boris solver code only. Should not interact with the existing solver classes.

### CODE REVIEW

File: python/boris_solver.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+        u_n = v0 * gamma_n
+        x_n = x0
+        v_n = v0        
+
+        # Half position step
+        x_nplushalf = x_n + u_n * time_step / (2.0 * gamma_n)
+
+        # Do the first half of the radiation reaction acceleration
+        u_minus = u_n + (time_step / 2.0) * self.radiation_acceleration(x_nplushalf, u_n)
+        gamma_minus = np.sqrt( 1 + (np.linalg.norm(u_n)/c)**2 )
```
Comment (TomGoffrey) [inline python/boris_solver.py]: Wouldn't `np.sqrt(1.0 + np.dot(u_n, u_n) / c**2)` be more efficient here?

File: python/boris_solver.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+        Returns the position and velocity at the next step
+        """
+
+        gamma_n = 1 / np.sqrt( 1 - (np.linalg.norm(v0)/c)**2 )    
+        u_n = v0 * gamma_n
+        x_n = x0
+        v_n = v0        
+
+        # Half position step
+        x_nplushalf = x_n + u_n * time_step / (2.0 * gamma_n)
```
Comment (TomGoffrey) [inline python/boris_solver.py]: Would be simpler to use ` v_n` directly rather than `u_n / gamma_n`

File: python/boris_solver.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+        t = np.tan(theta/2.0) * B_nplushalf / np.linalg.norm(B_nplushalf) 
+        u_prime = u_minus + np.cross(u_minus, t)
+        u_plus = u_minus + 2.0 / (1.0 + np.dot(t, t)) * np.cross(u_prime, t)
+
+        # Second half of the radiation reaction acceleration
+        u_nplus1 = u_plus + (time_step / 2.0) * self.radiation_acceleration(x_nplushalf, u_plus)
+
+        # Now update position
+        gamma_nplus1 = np.sqrt( 1 + (np.linalg.norm(u_nplus1)/c)**2 )
+        x_nplus1 = x_nplushalf + u_nplus1 * time_step / (2.0 * gamma_nplus1)
```
Comment (TomGoffrey) [inline python/boris_solver.py]: Similar to a previous comment, if you calculate `v_nplus1` first the update to `x_nplus1` is simpler.

File: python/boris_solver.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+        # Half position step
+        x_nplushalf = x_n + u_n * time_step / (2.0 * gamma_n)
+
+        # Do the first half of the radiation reaction acceleration
+        u_minus = u_n + (time_step / 2.0) * self.radiation_acceleration(x_nplushalf, u_n)
+        gamma_minus = np.sqrt( 1 + (np.linalg.norm(u_n)/c)**2 )
+
+        # Rotation step
+        B_nplushalf = np.zeros(3)
```
Comment (TomGoffrey) [inline python/boris_solver.py]: As `B_nplushalf` is set by the if statement this line could be removed.

File: python/boris_solver.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+        max_step = cfl / omega0
+
+        # Final time                                                                                
+        t_end = n_rotations * 2.0 * np.pi / omega0
+
+        # Calculate number of steps
+        n_steps = int(np.ceil(t_end / max_step))
+        step_size = t_end / n_steps
+
+        u_n = v0 * 1 / np.sqrt( 1 - (np.linalg.norm(v0)/c)**2 )
```
Comment (TomGoffrey) [inline python/boris_solver.py]: Similar comment to line 111.

File: python/boris_solver.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+
+        # Maximum time step                                                                          
+        max_step = cfl / omega0
+
+        # Final time                                                                                
+        t_end = n_rotations * 2.0 * np.pi / omega0
+
+        # Calculate number of steps
+        n_steps = int(np.ceil(t_end / max_step))
+        step_size = t_end / n_steps
```
Comment (TomGoffrey) [inline python/boris_solver.py]: Shouldn't this just be equal to `max_step` above?

**Review by TomGoffrey — COMMENTED** [review]
@sebj101 I've highlighted a few places where I think minor improvements could be made to this code. They're all rather minor optimisations, but worth considering. The code itself looks like it functions as intended.

**Comment by sebj101** [comment]
@TomGoffrey thanks for the feedback. I've implemented your recommendations.

**Review by TomGoffrey — APPROVED** [review]

**Pull request merged**